_posts/2018-01-09-dagger2-android-subcomponents.md

```diff
@@ -0,0 +1,205 @@
+
+##### 3. @Singleton
+Last but most interesting annotation in above code is `@Singleton`. This annotation specifies scope of the dependancy being provided. `Singleton` is most wide scope that can be used in `Dagger2`. A `Singleton` dependancy is initialised once **during `Component`'s lifecycle** in `Dagger2`.
+
+That's that, since now we have all the basics and our dependancies ready for injection lets go ahead and create our top level component called `AppComponent` in movie example.
+
+Here is the code for simplified version of `AppComponent`
+```java
+@Singleton
+@Component(modules = {AppModule.class, ApiModule.class})
+public interface AppComponent {
+  void inject(BaseApplication baseApplication);
+}
+```
+Lets have a look at new annotation introduced here
+
+##### @Component
+`@Component` tells dagger that following interface is a `Component` and you need to generate implementation for it. As soon as you write `@Component` and hit build in Android Studio, `Dagger2` generates implementation for the target interface. `Dagger2` adds a prefix `Dagger` in name of implementation class. For an example for our `AppComponent` dagger will generate a class called `DaggerAppComponent`. Also as it apparent from code `@Component` also takes modules which will construct and provide dependancies.
+
+### Get Set Inject!
+
+We have everything that we need, lets inject these dependencies so that we can actually use them. Since this is a global component, we will initialise it in `BaseApplication` class
+
+```java
+public class BaseApplication extends Application {
+    @Getter
+    private AppComponent appComponent;
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        appComponent = DaggerAppComponent.builder()
+                .appModule(new AppModule(this))
+                .build();
+
+        appComponent.inject(this);
+    }
+}
+```
+Most interesting piece of code is the one building dependancy graph. Note that even though `AppComponent` uses two modules only one is specified here, that is because `ApiModule` has no dependencies. Dagger can figure that out by itself and creates it using default constructor.
+
+### Where are my screens ?
+Now that we are done setting up global dependencies, about time we start creating screens. Our first screen is movie list screen. It follows standard MVP architecture where we have a `Fragment` inside and `Activity` as View, a `Presenter` and a `Contract` interface that defines language in which View talks to Presenter.
+
+Our view will thus need and instance of `Presenter`. `Presenter` is responsible for fetching data from API and providing it to View, so it will need `ApiService` and `View`. `Presenter` is where we will keep our business logic and it will not have any Android stuff so that we can unit test it easily.
```

Comment:
The point about Presenter not having any Android dependency is important. Should we stress on it more?